### PR TITLE
[682e] children_by_path is slow if called repeatedly

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftersoftware/craftercms/issues/682

### CODE REVIEW

File: ui/app/src/services/content.ts [hunk 1]
```diff
@@ -1195,47 +1195,33 @@ export function fetchChildrenByPath(
...
-        total,
-        offset,
-        limit
-      })
-    )
-  );
+  const paths = {
+    [path]: { ...options }
+  };
+  return fetchChildrenByPaths(siteId, paths).pipe(map((data) => data[path]));
```
Comment (rart) [inline ui/app/src/services/content.ts]:
```suggestion
  return fetchChildrenByPaths(siteId, { [path]: options }).pipe(pluck(path));
```

File: ui/app/src/services/content.ts [hunk 2]
```diff
@@ -1195,47 +1195,33 @@ export function fetchChildrenByPath(
...
-          throw error;
-        }
-      })
-    )
-  );
-  return forkJoin(requests).pipe(
-    map((responses) => {
+  return postJSON(`/studio/api/2/content/${siteId}/children`, { paths }).pipe(
+    pluck('response'),
+    map(({ items }) => {
```
Comment (rart) [inline ui/app/src/services/content.ts]:
```suggestion
    pluck('response', 'items'),
    map((items) => {
```

File: ui/app/src/services/content.ts [hunk 3]
```diff
@@ -1195,47 +1195,33 @@ export function fetchChildrenByPath(
...
   }
-  const requests = paths.map((path) =>
-    fetchChildrenByPath(siteId, path, { ...options, ...fetchOptionsByPath[path] }).pipe(
-      catchError((error: AjaxError) => {
-        if (error.status === 404) {
-          return of([]);
-        } else {
-          throw error;
-        }
-      })
```
Comment (rart) [inline ui/app/src/services/content.ts:1247]: What about this error handling? Anything to consider here?

**Review by rart — CHANGES_REQUESTED** [review]
These PRs are all active but this is not the whole task. All the navigators logic to use this is not done?

File: ui/app/src/state/actions/pathNavigator.ts [hunk 4]
```diff
@@ -67,6 +67,10 @@ export const pathNavigatorRefresh = /*#__PURE__*/ createAction<{ id: string }>('
 
 export const pathNavigatorBackgroundRefresh = /*#__PURE__*/ createAction<{ id: string }>('PATH_NAV_BACKGROUND_REFRESH');
 
+export const pathNavigatorsBackgroundRefresh = /*#__PURE__*/ createAction<{ ids: string[] }>(
+  'PATH_NAVS_BACKGROUND_REFRESH'
```
Comment (rart) [inline ui/app/src/state/actions/pathNavigator.ts]:
```suggestion
  'PATH_NAV_BULK_BG_REFRESH'
```

File: ui/app/src/state/actions/pathNavigatorTree.ts [hunk 5]
```diff
@@ -44,6 +44,10 @@ export const pathNavigatorTreeBackgroundRefresh = /*#__PURE__*/ createAction<{ i
   'PATH_NAV_TREE_BACKGROUND_REFRESH'
 );
 
+export const pathNavigatorTreesBackgroundRefresh = /*#__PURE__*/ createAction<{ ids: string[] }>(
+  'PATH_NAV_TREES_BACKGROUND_REFRESH'
```
Comment (rart) [inline ui/app/src/state/actions/pathNavigatorTree.ts]:
```suggestion
  'PATH_NAV_TREE_BULK_BG_REFRESH'
```

File: ui/app/src/state/actions/pathNavigator.ts (hunk 4 above)
Comment (rart) [reply ui/app/src/state/actions/pathNavigator.ts]: Rename const too (for all)

File: ui/app/src/services/content.ts (hunk 1 above)
Comment (jvega190) [reply ui/app/src/services/content.ts]: pluck is deprecated, should we still use it?

Comment (rart) [reply ui/app/src/services/content.ts]: no

File: ui/app/src/services/content.ts [hunk 6]
```diff
@@ -1195,47 +1195,30 @@ export function fetchChildrenByPath(
...
-    )
-  );
-  return forkJoin(requests).pipe(
-    map((responses) => {
+  return postJSON(`/studio/api/2/content/${siteId}/children`, { paths }).pipe(
+    pluck('response', 'items'),
+    map((items) => {
```
Comment (rart) [inline ui/app/src/services/content.ts]: If pluck is deprecated, then let's begin progressively removing wherever it is found.

**Review by rart — CHANGES_REQUESTED** [review]

**Review by rart — CHANGES_REQUESTED** [review]
It seems we can still bulk more on some events. Anywhere where using the `actions.push(...)` pattern. Ping me to review if necessary.

File: ui/app/src/state/epics/pathNavigator.ts [hunk 7]
```diff
@@ -126,6 +130,58 @@ export default [
...
+          map(([items, children]) => {
+            const actions = [];
+            ids.forEach((id) => {
+              actions.push(
+                pathNavigatorFetchPathComplete({
+                  id,
+                  parent: items.find((item) => item.path.startsWith(withoutIndex(state.pathNavigator[id].currentPath))),
+                  children: children[state.pathNavigator[id].currentPath]
+                })
+              );
```
Comment (rart) [inline ui/app/src/state/epics/pathNavigator.ts]: Create an action that receives the bulk response and is processed accordingly in the reducer. Follow the action, actionComplete, actionFailed pattern.

File: ui/app/src/state/epics/pathNavigator.ts [hunk 8]
```diff
@@ -375,15 +431,15 @@ export default [
...
+            idsToRefresh.push(navigator.id);
           } /* else if (
             // Case (c) - Content epics load any item that's on the state already
             navigator.currentPath === getParentPath(parentPathOfTargetPath)
           ) {
             actions.push(fetchSandboxItem({ path: parentPathOfTargetPath }));
           } */
         });
-        return actions;
+        return of(pathNavigatorBulkBackgroundRefresh({ ids: idsToRefresh }));
```
Comment (rart) [inline ui/app/src/state/epics/pathNavigator.ts]: If we're returning a single action, you can switch from `mergeMap` to `map` and dismiss the use of `of`.
Do not dispatch if ids is empty

File: ui/app/src/state/epics/pathNavigator.ts [hunk 9]
```diff
@@ -418,18 +474,22 @@ export default [
...
           } else if (
             withoutIndex(navigator.currentPath) === parentOfTargetPath ||
             withoutIndex(navigator.currentPath) === parentOfSourcePath
           ) {
-            actions.push(pathNavigatorBackgroundRefresh({ id: navigator.id }));
+            idsToBgRefresh.push(navigator.id);
           }
         });
+        actions.push(pathNavigatorBulkRefresh({ ids: idsToRefresh }));
+        actions.push(pathNavigatorBulkBackgroundRefresh({ ids: idsToBgRefresh }));
```
Comment (rart) [inline ui/app/src/state/epics/pathNavigator.ts]:
```suggestion
        idsToRefresh.length && actions.push(pathNavigatorBulkRefresh({ ids: idsToRefresh }));
        idsToBgRefresh.length && actions.push(pathNavigatorBulkBackgroundRefresh({ ids: idsToBgRefresh }));
```

File: ui/app/src/state/epics/pathNavigator.ts [hunk 10]
```diff
@@ -441,13 +501,13 @@ export default [
...
-        const actions = [];
+        const ids = [];
         Object.values(state.pathNavigator).forEach((tree) => {
           if (['/templates', '/scripts', '/static-assets'].includes(getRootPath(tree.rootPath))) {
-            actions.push(pathNavigatorBackgroundRefresh({ id: tree.id }));
+            ids.push(tree.id);
           }
         });
-        return actions;
+        return of(pathNavigatorBulkBackgroundRefresh({ ids }));
```
Comment (rart) [inline ui/app/src/state/epics/pathNavigator.ts]: mergeMap => map, no of
do not dispatch if ids is empty.

**Review by rart — CHANGES_REQUESTED** [review]
Some of the revisions the comments apply to many places but I didn't write the same comment everywhere (e.g. don't dispatch if the arrays of ids are empty, don't use single actionComplete handlers to dispatch bulk action results, etc). Check all the work for those and apply everywhere.

File: ui/app/src/state/epics/pathNavigator.ts [hunk 11]
```diff
@@ -440,14 +495,14 @@ export default [
...
+      map(([, state]) => {
+        const ids = [];
         Object.values(state.pathNavigator).forEach((tree) => {
           if (['/templates', '/scripts', '/static-assets'].includes(getRootPath(tree.rootPath))) {
-            actions.push(pathNavigatorBackgroundRefresh({ id: tree.id }));
+            ids.push(tree.id);
           }
         });
-        return actions;
+        return ids.length ? pathNavigatorBulkBackgroundRefresh({ ids }) : NEVER;
```
Comment (rart) [inline ui/app/src/state/epics/pathNavigator.ts]: Can't return an observable (NEVER) from map. This is obviously untested... 👎🏻

File: ui/app/src/state/epics/pathNavigatorTree.ts [hunk 12]
```diff
@@ -308,13 +441,16 @@ export default [
             ) {
               const pathToUpdate = parentPath in tree.totalByPath ? parentPath : withIndex(parentPath);
               // Show the new child
-              pathToUpdate in tree.childrenByParentPath &&
-                actions.push(pathNavigatorTreeFetchPathChildren({ id, path: pathToUpdate, expand: false }));
+              if (pathToUpdate in tree.childrenByParentPath) {
+                fetchPathsData.push({ id, path: pathToUpdate, expand: false });
+              }
```
Comment (rart) [inline ui/app/src/state/epics/pathNavigatorTree.ts]: What's this for? Doesn't seem to have any usage?
This code path is getting lost from what it did before...

File: ui/app/src/state/epics/pathNavigator.ts [hunk 13]
```diff
@@ -126,6 +130,52 @@ export default [
...
+          ? forkJoin([
+              fetchItemsByPath(state.sites.active, paths, { castAsDetailedItem: true }),
+              fetchChildrenByPaths(state.sites.active, optionsByPath)
+            ]).pipe(
+              map(([items, children]) =>
+                pathNavigatorBulkFetchPathComplete({
+                  paths: ids.map(({ id }) => ({
+                    id,
+                    parent: items.find((item) =>
+                      item.path.startsWith(withoutIndex(state.pathNavigator[id].currentPath))
```
Comment (jvega190) [inline ui/app/src/state/epics/pathNavigator.ts:167]: The backend doesn't send back the children in the same order as the items

File: ui/app/src/state/epics/pathNavigatorTree.ts [hunk 14]
```diff
@@ -166,8 +181,90 @@ export default [
       })
     ),
   // endregion
-  // region pathNavigatorFetchPathChildren
+  // region pathNavigatorTreeBulkBackgroundRefresh
   (action$, state$) =>
+    action$.pipe(
+      // TODO: There's no special handling for background vs visible. Should there be?
```
Comment (jvega190) [inline ui/app/src/state/epics/pathNavigatorTree.ts:187]: There doesn't seem to be any previous handling for background vs visible (in comparison with pathNavigator, that is in the reducer). The pathNavTree state doesn't have any `isFetching` or something implemented for the loading state to display. Should we create a ticket for that?

**Review by rart — CHANGES_REQUESTED** [review]
Apply changes from 4.1.x

**Review by rart — APPROVED** [review]

**Pull request merged**